File: ios/Sources/NativeLoaderPlugin/NativeLoaderPlugin.swift
import Foundation
import Capacitor

@objc(NativeLoaderPlugin)
public class NativeLoaderPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NativeLoaderPlugin"
    public let jsName = "NativeLoader"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "configure", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "show", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "update", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setProgress", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hide", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hideAll", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setWebViewLayout", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resetWebViewLayout", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = NativeLoader.shared

    @objc func configure(_ call: CAPPluginCall) {
        let defaults = call.getObject("defaults") ?? [:]
        implementation.configure(defaults: defaults)
        call.resolve()
    }

    @objc func show(_ call: CAPPluginCall) {
        let id = implementation.show(options: stringOptions(call.options), webView: bridge?.webView)
        call.resolve(["id": id])
    }

    @objc func update(_ call: CAPPluginCall) {
        guard call.getString("id") != nil else {
            call.reject("Missing required parameter: id")
            return
        }
        implementation.update(options: stringOptions(call.options))
        call.resolve()
    }

    @objc func setProgress(_ call: CAPPluginCall) {
        guard let progress = call.getDouble("progress") else {
            call.reject("Missing required parameter: progress")
            return
        }
        implementation.setProgress(id: call.getString("id"), progress: progress)
        call.resolve()
    }

    @objc func hide(_ call: CAPPluginCall) {
        implementation.hide(
            id: call.getString("id"),
            animated: call.getBool("animated") ?? true,
            restoreWebView: call.getBool("restoreWebView") ?? true
        )
        call.resolve()
    }

    @objc func hideAll(_ call: CAPPluginCall) {
        implementation.hideAll(
            animated: call.getBool("animated") ?? true,
            restoreWebView: call.getBool("restoreWebView") ?? true
        )
        call.resolve()
    }

    @objc func setWebViewLayout(_ call: CAPPluginCall) {
        implementation.setWebViewLayout(options: stringOptions(call.options), webView: bridge?.webView)
        call.resolve()
    }

    @objc func resetWebViewLayout(_ call: CAPPluginCall) {
        implementation.resetWebViewLayout(animated: call.getBool("animated") ?? true)
        call.resolve()
    }

    @objc func getState(_ call: CAPPluginCall) {
        let state = implementation.getState()
        call.resolve([
            "showing": state.0,
            "ids": state.1
        ])
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve([
            "version": implementation.getPluginVersion()
        ])
    }
}

private func stringOptions(_ options: [AnyHashable: Any]) -> [String: Any] {
    return options.reduce(into: [String: Any]()) { result, entry in
        guard let key = entry.key as? String else { return }
        result[key] = entry.value
    }
}

File: android/src/main/kotlin/app/capgo/nativeloader/NativeLoaderPlugin.kt
package app.capgo.nativeloader

import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import org.json.JSONArray
import org.json.JSONObject

@CapacitorPlugin(name = "NativeLoader")
class NativeLoaderPlugin : Plugin() {

    @PluginMethod
    fun configure(call: PluginCall) {
        val defaults = call.getObject("defaults")?.toMap() ?: emptyMap()
        NativeLoader.configure(defaults)
        call.resolve()
    }

    @PluginMethod
    fun show(call: PluginCall) {
        val id = NativeLoader.show(activity, call.data.toMap(), bridge.webView)
        call.resolve(JSObject().put("id", id))
    }

    @PluginMethod
    fun update(call: PluginCall) {
        if (call.getString("id").isNullOrBlank()) {
            call.reject("Missing required parameter: id")
            return
        }
        NativeLoader.update(call.data.toMap())
        call.resolve()
    }

    @PluginMethod
    fun setProgress(call: PluginCall) {
        val progress = call.getDouble("progress")
        if (progress == null) {
            call.reject("Missing required parameter: progress")
            return
        }
        NativeLoader.setProgress(call.getString("id"), progress)
        call.resolve()
    }

    @PluginMethod
    fun hide(call: PluginCall) {
        NativeLoader.hide(
            id = call.getString("id"),
            animated = call.getBoolean("animated") ?: true,
            restoreWebView = call.getBoolean("restoreWebView") ?: true,
        )
        call.resolve()
    }

    @PluginMethod
    fun hideAll(call: PluginCall) {
        NativeLoader.hideAll(
            animated = call.getBoolean("animated") ?: true,
            restoreWebView = call.getBoolean("restoreWebView") ?: true,
        )
        call.resolve()
    }

    @PluginMethod
    fun setWebViewLayout(call: PluginCall) {
        NativeLoader.setWebViewLayout(call.data.toMap(), bridge.webView)
        call.resolve()
    }

    @PluginMethod
    fun resetWebViewLayout(call: PluginCall) {
        NativeLoader.resetWebViewLayout(call.getBoolean("animated") ?: true)
        call.resolve()
    }

    @PluginMethod
    fun getState(call: PluginCall) {
        val state = NativeLoader.getState()
        call.resolve(JSObject().put("showing", state.first).put("ids", JSArray(state.second)))
    }

    @PluginMethod
    fun getPluginVersion(call: PluginCall) {
        call.resolve(JSObject().put("version", NativeLoader.getPluginVersion()))
    }

    override fun handleOnDestroy() {
        NativeLoader.hideAll(animated = false, restoreWebView = true)
        super.handleOnDestroy()
    }
}

private fun JSONObject.toMap(): Map<String, Any?> {
    val output = mutableMapOf<String, Any?>()
    keys().forEach { key ->
        output[key] = normalizeJsonValue(opt(key))
    }
    return output
}

private fun JSONArray.toListValue(): List<Any?> {
    return (0 until length()).map { index ->
        normalizeJsonValue(opt(index))
    }
}

private fun normalizeJsonValue(value: Any?): Any? {
    return when (value) {
        JSONObject.NULL -> null
        is JSONObject -> value.toMap()
        is JSONArray -> value.toListValue()
        else -> value
    }
}
